spec/requests/uploads_controller_spec.rb

```diff
@@ -774,7 +778,577 @@ def upload_file(file, folder = "images")
+            upload_type: "composer",
+            content_type: "image/png",
+            file_size: 1024
+          }
+          post "/uploads/create-multipart.json", params: {
+            file_name: "test.png",
+            upload_type: "composer",
+            content_type: "image/png",
+            file_size: 1024
+          }
+        end
+        expect(response.status).to eq(429)
+      end
+    end
+
+    context "when the store is not external" do
+      before do
+        sign_in(user)
+      end
+
+      it "returns 404" do
+        post "/uploads/create-multipart.json", params: {
+          file_name: "test.png",
+          upload_type: "composer",
+          content_type: "image/png",
+          file_size: 1024
+        }
+        expect(response.status).to eq(404)
+      end
+    end
+  end
+
+  describe "#batch_presign_multipart_parts" do
+    let(:mock_multipart_upload_id) { "ibZBv_75gd9r8lH_gqXatLdxMVpAlj6CFTR.OwyF3953YdwbcQnMA2BLGn8Lx12fQNICtMw5KyteFeHw.Sjng--" }
+    let!(:external_upload_stub) do
+      Fabricate(:image_external_upload_stub, created_by: user, multipart: true, external_upload_identifier: mock_multipart_upload_id)
+    end
+
+    context "when the store is external" do
+      before do
+        sign_in(user)
+        SiteSetting.enable_direct_s3_uploads = true
+        setup_s3
+      end
+
+      def stub_list_multipart_request
+        list_multipart_result = <<~BODY
+        <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n
+        <ListPartsResult>
+           <Bucket>s3-upload-bucket</Bucket>
+           <Key>#{external_upload_stub.key}</Key>
+           <UploadId>#{mock_multipart_upload_id}</UploadId>
+           <PartNumberMarker>0</PartNumberMarker>
+           <NextPartNumberMarker>0</NextPartNumberMarker>
+           <MaxParts>1</MaxParts>
+           <IsTruncated>false</IsTruncated>
+           <Part>
+              <ETag>test</ETag>
+              <LastModified>#{Time.zone.now}</LastModified>
+              <PartNumber>1</PartNumber>
+              <Size>#{5.megabytes}</Size>
+           </Part>
+           <Initiator>
+              <DisplayName>test-upload-user</DisplayName>
+              <ID>arn:aws:iam::123:user/test-upload-user</ID>
+           </Initiator>
+           <Owner>
+              <DisplayName></DisplayName>
+              <ID>12345</ID>
+           </Owner>
+           <StorageClass>STANDARD</StorageClass>
+        </ListPartsResult>
+        BODY
+        stub_request(:get, "https://s3-upload-bucket.s3.us-west-1.amazonaws.com/#{external_upload_stub.key}?uploadId=#{mock_multipart_upload_id}").to_return({ status: 200, body: list_multipart_result })
+      end
+
+      it "errors if the correct params are not provided" do
+        post "/uploads/batch-presign-multipart-parts.json", params: {}
+        expect(response.status).to eq(400)
+      end
+
+      it "errors if the part_numbers do not contain numbers between 1 and 10000" do
+        stub_list_multipart_request
```

Comment:
Do we need to stop the request here when it wouldn't be called since stuff will error out before that request happens?

Reply:
We actually do the `list` request before we check part numbers, but it makes more sense to check part numbers first, so I will switch the order around :+1: